tests/integration/devfile/cmd_devfile_app_test.go

```diff
@@ -25,89 +25,162 @@ var _ = Describe("odo devfile app command tests", func() {
+	When("the user creates and pushes two new devfile components in different apps", func() {
+		var context0 string
+		var context1 string
+		var component0 string
+		var component1 string
+
+		var app0 string
+		var app1 string
+
+		BeforeEach(func() {
+			app0 = helper.RandString(4)
+			app1 = helper.RandString(4)
+
+			// create first component in the first app
+			context0 = helper.CreateNewContext()
+			component0 = helper.RandString(4)
+			helper.Cmd("odo", "create", "nodejs", "--project", namespace, component0, "--context", context0, "--app", app0).ShouldPass()
+			helper.CopyExample(filepath.Join("source", "devfiles", "nodejs", "project"), context0)
+			helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile.yaml"), filepath.Join(context0, "devfile.yaml"))
+			helper.Cmd("odo", "push", "--context", context0).ShouldPass()
+
+			// create second component in the second app
+			context1 = helper.CreateNewContext()
+			component1 = helper.RandString(4)
+			helper.Cmd("odo", "create", "nodejs", "--project", namespace, component1, "--context", context1, "--app", app1).ShouldPass()
+			helper.CopyExample(filepath.Join("source", "devfiles", "nodejs", "project"), context1)
+			helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile.yaml"), filepath.Join(context1, "devfile.yaml"))
+			helper.Cmd("odo", "push", "--context", context1).ShouldPass()
 		})
-	})
 
-	Context("Testing URLs for OpenShift specific scenarios", func() {
-		JustBeforeEach(func() {
-			if os.Getenv("KUBERNETES") == "true" {
-				Skip("This is a OpenShift specific scenario, skipping")
-			}
+		When("the user creates and pushes a third s2i component on a openshift cluster", func() {
+
+			var context00 string
+			var component00 string
+
+			var storage00 string
+			var url00 string
+
+			BeforeEach(func() {
+				if os.Getenv("KUBERNETES") == "true" {
+					Skip("This is a OpenShift specific scenario, skipping")
+				}
+
+				context00 = helper.CreateNewContext()
+				component00 = helper.RandString(4)
+				storage00 = helper.RandString(4)
+				url00 = helper.RandString(4)
+				helper.CopyExample(filepath.Join("source", "nodejs"), context00)
+				helper.Cmd("odo", "component", "create", "--s2i", "nodejs", component00, "--app", app0, "--project", namespace, "--context", context00).ShouldPass()
+				helper.Cmd("odo", "storage", "create", storage00, "--path", "/data", "--size", "1Gi", "--context", context00).ShouldPass()
+				helper.Cmd("odo", "url", "create", url00, "--port", "8080", "--context", context00).ShouldPass()
+				helper.Cmd("odo", "push", "--context", context00).ShouldPass()
+			})
+
+			It("should list, describe and delete the app properly with json output", func() {
+				runner(testInfo{
+					app0:      app0,
+					app1:      app1,
+					comp0:     component0,
+					comp00:    component00,
+					url00:     url00,
+					storage00: storage00,
+					namespace: namespace,
+				})
+			})
+
+			AfterEach(func() {
```

Comment:
I find easier to maintain if AfterEach is placed just after the BeforeEach, so we can see what is created / freed more easily (similar to defer Close() is written just after Open()).